BreathData.cpp

```diff
@@ -0,0 +1,146 @@
+  display.writeBPM(60000.0/cycleDuration);             // measured respiratory rate (seconds)
+  display.writeO2(O2);                                 // measured FIO2 concentration (@FutureWork: pending addition of O2 sensor)  
+
+  // close expiratory valve
+  expValve.close();
+
+  // Compute intervals at current settings
+  unsigned long targetCycleDuration = 60000UL / display.bpm(); // ms from start of cycle to end of inspiration
+  targetInspDuration = 105 * targetCycleDuration * display.inspPercent() / 10000; // allowing a bit more time to complete tidal volume inhailation
+  targetCycleEndTime = cycleTimer + targetCycleDuration;                          // target time for breath to end (for HOLD_EXP_STATE to end)
+  targetInspEndTime  = cycleTimer + targetInspDuration;                           // target time for INSP_STATE to end
+  targetExpDuration  = targetCycleDuration - targetInspDuration - MIN_PEEP_PAUSE; // target time for EXP_STATE to end
+  desiredInspFlow = display.volume() * CC_PER_MS_TO_LPM / targetInspDuration;     // desired inspiratory flowrate cc/ms
+
+  // begin PID control based on desired flow and reset tidal volume
+  inspValve.beginBreath(desiredInspFlow); 
+  inspFlowReader.resetVolume();           
+  cycleCount++;                           
+}
+
+
+/** 
+ * function to check if sensor readings are within acceptable ranges
+ * activate an alarm if they are not and deactivate the alarm once they are again
+ * 
+ * @param reading -- sensor reading
+ * @param compareValue -- value to compare reading to
+ * @param sensitivity -- the range of error
+ * @param highAlarmCode -- the high alarm code for a specific sensor
+ * @param lowAlarmCode -- the low alarm code for a specific sensor
+ */ 
+void BreathData::checkAlarmRangeWithUpdate(float reading, float &compareValue, float sensitivity, alarmCode highAlarmCode, alarmCode lowAlarmCode) { 
+  // if not the first reading, compare and alarm if abnormal
+  if (!isnan(compareValue)) {
+    if (reading > compareValue + sensitivity) { 
+      alarmMgr.activateAlarm(highAlarmCode);
+    } else if (reading < compareValue - sensitivity) { 
+      alarmMgr.activateAlarm(lowAlarmCode);
+    } else {
+      alarmMgr.deactivateAlarm(lowAlarmCode); 
+      compareValue = reading; // update to remember value for next comparison
+    }
+  } else {
+    compareValue = reading; // update to remember value for next comparison
+  }
+}
+
+/** 
+ * check against a value that doesn't need to be stored -- call above with dummy variable
+ */ 
+void BreathData::checkAlarmRange(float reading, float compareValue, float sensitivity, alarmCode highAlarmCode, alarmCode lowAlarmCode){ 
+  float dummyCompareValue = compareValue;
+  checkAlarmRangeWithUpdate(reading, dummyCompareValue, sensitivity, highAlarmCode, lowAlarmCode);
+}
+
+/** 
+ * check for errors in sensor readings and take appropriate action
+ */ 
+void BreathData::checkSensorReadings() {
```

Comment:
Unclear why this is in this class instead of being in the sensors module.

Reply:
We don't really have a sensors module yet.  This is here at the moment because it checks whether the current breath values are in alarm ranges.  Probably the name of the function should be changed to checkForAlarmConditions.